htlcswitch/pending_payment.go

```diff
@@ -0,0 +1,492 @@
+// payment ID. If the pid is not found, ErrPaymentIDNotFound will be returned.
+func (store *pendingPaymentStore) getPayment(pid uint64) (*PendingPayment,
+	error) {
+
+	var paymentIDBytes [8]byte
+	binary.BigEndian.PutUint64(paymentIDBytes[:], pid)
+
+	var p *PendingPayment
+	err := store.db.View(func(tx *bbolt.Tx) error {
+		pendingPayments := tx.Bucket(pendingPaymentBucketKey)
+		if pendingPayments == nil {
+			return ErrPaymentIDNotFound
+		}
+
+		bucket := pendingPayments.Bucket(paymentIDBytes[:])
+		if bucket == nil {
+			return ErrPaymentIDNotFound
+		}
+
+		pendingBytes := bucket.Get(pendingPaymentKey)
+		if pendingBytes == nil {
+			return ErrPaymentIDNotFound
+		}
+
+		var err error
+		r := bytes.NewReader(pendingBytes)
+		p, err = deserializePendingPayment(r)
+		if err != nil {
+			return err
+		}
+
+		return nil
+
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// getPaymentResult is used to get the payment result for the given payment ID.
+// It returns a channel on which the result will be delivered when ready. If
+// the payment ID is not found in the store, ErrPaymentIDNotFound is returned.
+func (store *pendingPaymentStore) getPaymentResult(pid uint64) (
+	<-chan PaymentResult, error) {
+
+	var paymentIDBytes [8]byte
+	binary.BigEndian.PutUint64(paymentIDBytes[:], pid)
+
+	resultChan := make(chan PaymentResult, 1)
+	err := store.db.View(func(tx *bbolt.Tx) error {
+		pendingPayments := tx.Bucket(pendingPaymentBucketKey)
+		if pendingPayments == nil {
+			return ErrPaymentIDNotFound
+		}
+
+		bucket := pendingPayments.Bucket(paymentIDBytes[:])
+		if bucket == nil {
+			return ErrPaymentIDNotFound
+		}
+
+		// Check whether a result is already available.
+		var result PaymentResult = &PaymentSuccess{}
+		resultBytes := bucket.Get(successKey)
+		if resultBytes == nil {
+			// If no success result was available, check for
+			// failure.
+			result = &PaymentFailure{}
+			resultBytes = bucket.Get(failureKey)
+		}
+
+		if resultBytes == nil {
+			// Result not yet available, we will notify once a
+			// result is available.
+			// NOTE: we do this inside the db transaction, to avoid
+			// a race where the payment result is written to the db
+			// between this db tx and we add the resultChan to the
+			// map.
+			store.resultsMtx.Lock()
+			store.results[pid] = append(
+				store.results[pid], resultChan,
+			)
+			store.resultsMtx.Unlock()
+			return nil
+		}
+
+		// Decode the result we found, and send it on the result
+		// channel.
+		r := bytes.NewReader(resultBytes)
+		if err := result.Decode(r); err != nil {
+			return err
+		}
+
+		resultChan <- result
+		return nil
+
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return resultChan, nil
+}
+
+// fetchPendingPayments retrieves all payments from the store that are pending,
+// meaning no results are available yet, and can be re-forwarded on the
+// network.
+func (store *pendingPaymentStore) fetchPendingPayments() ([]*PendingPayment,
```

Comment:
Looks like this is test only code. Would look for a different way to assert behaviour (intercepting mock?)